clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4097,11 +4097,12 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
   }
 
   void addArrayElementInit(FieldDecl *FD, QualType T) {
+    Expr *ArrParam = ArrayParamBases.pop_back_val();
     Expr *RCE = createReinterpretCastExpr(
-        createGetAddressOf(ArrayParamBases.pop_back_val()),
+        createGetAddressOf(ArrParam),
         SemaSYCLRef.getASTContext().getPointerType(T));
     Expr *Initializer = createDerefOp(RCE);
-    addFieldInit(FD, T, Initializer);
+    addFieldInit(FD, T, (T->isPointerType()) ? ArrParam : Initializer);
```

Comment:
Can we not create AST nodes that we are not going to use?
```suggestion
    Expr *Initializer = ArrayParamBases.pop_back_val();
    if (!T->isPointerType()) {
      Expr *RCE = createReinterpretCastExpr(
        createGetAddressOf(Initializer),
        SemaSYCLRef.getASTContext().getPointerType(T));
        Initializer = createDerefOp(RCE);
    }
    addFieldInit(FD, T, Initializer);
```
(formatting is off)